### PR TITLE
Add Variational Quantum Algorithms Module

### PR DESCRIPTION
**Pull request by EnBr55**
Functionality as outlined in #118

- [X] Core functionality
- [X] Following style requirements
- [X] API documentation
- [x] Tests
  - Mostly done, still in progress
- [x] Manual documentation
  - In progress

This PR introduces a new module, `qutip_qip.vqa`, for defining and simulating Variational Quantum Algorithms. For examples for how this module could be used, please see https://github.com/EnBr55/qutip-vqa-examples/

### CODE REVIEW

File: doc/source/apidoc/qutip_qip.vqa.rst
Comment (nathanshammah) [inline doc/source/apidoc/qutip_qip.vqa.rst:21]:
```suggestion
      VQABlock
```

Comment (EnBr55) [reply doc/source/apidoc/qutip_qip.vqa.rst:21]: Good catch, thanks.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:204]: In the new black styling released a few days ago, it is recommended to omit the space around `**`. That's why the lint test is failing. You will get it right with an updated version of black.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:821]: Same lint issue.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:10]: `matplotlib` is only an optional package required, so the test actions do not include that automatically. You can wrap this with a try-catch syntax to avoid failing tests or move this importation into the plotting function.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:737]: Better warp `|x_0x_1...x_n>` with the math mode, ``:math:`|x_0x_1...x_n>` ``. Underscores have special meanings in rst files, so the building doc test will complain.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:762]: Same with math ``:math:`|010>` ``

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:812]: Math mode as well

**Comment by BoxiLi** [comment]
Bingo! You passed most of the tests. The coverage report is still complaining but I understand more tests are on the way. You can click the button `Details` to see which lines are not covered.

You can also test the plotting function by simply running the function without checking any result. This does not check if the plot generated is correct but at least verifies that the workflow throws no errors. Like in 
https://github.com/qutip/qutip-qip/blob/860856cd2dbb24affe8795dd177ee14b32442ad0/tests/test_processor.py#L136-L153

**Comment by EnBr55** [comment]
Hey @BoxiLi where would be the most appropriate place for the doc page for this module? I was thinking the best match currently might be under https://qutip-qip.readthedocs.io/en/latest/qip-simulator.html but wasn't sure if it should have its own page instead? Thanks :)

**Comment by BoxiLi** [comment]
Yeah you can start a new page for it. Maybe also give a brief intro for VQA.

**Comment by BoxiLi** [comment]
It is a bit annoying that the test needs to be approved every time when a new push is made. Not sure if there is a way to just allow any future tests to run for this PR. Sorry for all the delay, @EnBr55 Ben, please feel free to drop me an email next time you add something and want to run the test. Sometimes I miss the GitHub email and am unaware of any new push.

We can also try to get this PR merged first and add new patches in a different one so that any future PR of yours won't need approval.

The doc test is failing because the online doc generator does not have latex installed. You can render a png locally and add it as a figure, just like the other part of doc.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:278]: This is a bit confusing, I understand that for debugging you might just want to evaluate the parameters. However, having this in the optimization method seems contradictory to the name of the method.
Is this necessary? Why not just call the evaluation function in debugging?

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:319]: This is only used if `layer_by_layer==False`

File: doc/source/qip-vqa.rst
Comment (BoxiLi) [inline doc/source/qip-vqa.rst:6]: You can add your name and/or your email to get some credit if you want. ;)

File: doc/source/qip-vqa.rst
Comment (BoxiLi) [inline doc/source/qip-vqa.rst:117]: You don't have to import his figure. Using the plot derivative you can directly show the matplotlib figure you generated from the code (https://qutip-qip.readthedocs.io/en/latest/contribution-docs.html#id4). Sorry if my previous comment misleads you. I meant only the circuit figure.

**Review by BoxiLi — COMMENTED** [review]
@EnBr55 This is just part of the review. The code looks very neat. Mostly just naming plus a few of my questions, to clarify things. You don't have to accept all of them and feel free to discuss :)

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: Agreed, thanks!

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py:163]: You're right - I think it's sensible to assume the initial state is 0, and the user can prepare a different initial state by adding `VQABlock`s with the "initial" flag set to true. Good call on the `basis([2] * n, [0] * n)`, that's much more convenient! 

I think it might be a bit more friendly to the user if they can call `VQA.get_initial_state()` rather than having to use `basis([2] * VQA.n_qubits, [0] * VQA.n_qubits)`, even if the function is that simple? Happy to hear your thoughts, though.

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: Oops, I think I meant to remove this! Has been removed now.

File: doc/source/qip-vqa.rst
Comment (EnBr55) [reply doc/source/qip-vqa.rst]: Got it, thanks! I've switched to using the plot directive for this section now.

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: Thanks, I've modified the all the messaging that specifies `self` with something more clear.

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: I've updated this to use bullet points now. Is it correct to space things with newlines like this within the docstring?

```python
"""
[...]
conditions for the optimizer. Otherwise if a string is given:

* (Default) "random" will randomize initial free parameters
  between 0 and 1.

* "ones" will set each initial free parameter to a value of 1.
"""
```

Comment (BoxiLi) [reply src/qutip_qip/vqa.py]: Seems to work. 
image

You can also download it in our GitHub action 
"Build HTML documentation" from the generated artifacts.(https://github.com/apps/github-actions)@github-actions](https://github.com/qutip/qutip-qip/actions/runs/2201435768)

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: Good catch, thanks.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:697]: This part has some overlap with the `get_unitary_derivative` method. The method `get_unitary_frechet_derivative` is only used if the operator is `ParameterizedHamiltonian`.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:299]: Add to the docstring
```python
"""
Returns
-------
result: :obj:`.OptimizationResult`
    The optimized angles and final state.
"""
```

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:294]: Using Sphinx enumerating/listing?

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:506]: Is this the same as `num_parameters`? Better be consistent with the name.

File: doc/source/qip-vqa.rst
Comment (BoxiLi) [inline doc/source/qip-vqa.rst:46]:
```suggestion
After constructing this instance, we are ready to begin adding elements to our parameterized circuit. Circuit elements in this module are represented by :class:`.VQABlock` instances. Fundamentally, the role of this class is to generate an operator for the circuit. To do this, it keeps track of its free parameters, and gives information to the :class:`.VQA` instance on how to update them. The operator itself can be generated by a user-defined function call, a parameterized Hamiltonian to exponentiate, a pre-computed unitary operator, or a string referring to a gate that has already been defined (either by the user already, or a gate native to QuTiP).
```

File: doc/source/qip-vqa.rst
Comment (BoxiLi) [inline doc/source/qip-vqa.rst:48]: If I understand correctly
```suggestion
In the absence of specification, a VQA block takes :class:`~.Qobj` a Hamiltonian :math:`H`, and will generate a unitary with free parameter, :math:`\theta`, as :math:`U(\theta) = e^{-i \theta H}`. For example, 
```

It would be nice if examples for other types of VQA blocks can also be added. Can leave it for another PR.

File: doc/source/qip-vqa.rst
Comment (BoxiLi) [inline doc/source/qip-vqa.rst:113]:
```suggestion
  Accessing ``result.res.x``, we have the array of parameters found during optimization. In our case, we only had one free parameter, so we examine the first element of this array.
```

**Review by BoxiLi — CHANGES_REQUESTED** [review]
@EnBr55 Here is the second half of the review. Sorry for the delay.

- There are two png figures with the same content. One is duplication I guess (doc/circ.png)?

- If it is not too late, can we rename the public API
`n_qubits` -> `num_qubits`
`n_layers` -> `num_layers`
`n_paramters` -> `num_parameters`
to be consistent with the rest of the naming convention in qutip-qip? For the intermediate result not accessible to the user it does not matter.

- It is fine to merge as is. If you want to develop further on this module. I would recommend to define different types of VQA blocks as subclasses. In that way you don't have to use lots of if-else branches to check the input. You only need to specify this when defining the VQA blocks, e.g. `UnitaryVQABLock`, `HamiltonianVQABlock` etc..

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [reply src/qutip_qip/vqa.py:163]: Sure yeah it is a bit easier.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [reply src/qutip_qip/vqa.py:529]: I see that you already know this and used it in the rest doc. It also works for the API docs. `:obj:` here is the same as `:class:`.

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: Good call, I agree. I've added a check on the output when the function is called, just to avoid the possibility of weird behaviour if things don't entirely crash on non-`qobj` inputs somewhere.

File: src/qutip_qip/vqa.py
Comment (EnBr55) [reply src/qutip_qip/vqa.py]: You're right, sounds good to me. Changed and commented to make it a bit more clear.

File: src/qutip_qip/vqa.py
Comment (BoxiLi) [inline src/qutip_qip/vqa.py:711]: This seems the only issue that remained. I think `get_unitary_frechet_derivative` can be removed and the single part that is used
```python
            arg = -1j * self.operator.get_hamiltonian(angles)
            direction = -1j * self.operator.p_terms[term_index]
```
can be moved directly to `get_unitary_derivative`?

Once this is cleaned. I believe we are ready to merge!

Comment (EnBr55) [reply src/qutip_qip/vqa.py:711]: Was just writing a comment here haha. I'll post below anyway

That's true. I was originally thinking maybe the user would want to define a custom function to generate the Jacobian during optimization, which could depend on this method. Technically the method would have output the same as `get_unitary_derivative` if the operator is a `Qobj` Hamiltonian, but it's probably sensible to require that this method should only be called for `VQABlock`s that use a `ParameterizedHamiltonian`, so we could merge the two methods.

Comment (BoxiLi) [reply src/qutip_qip/vqa.py]: > I was originally thinking maybe the user would want to define a custom function to generate the Jacobian during optimization, which could depend on this method.

I see. Yes, custom derivatives could be useful. They can even overwrite the full `get_unitary_derivative` I guess? One can surely further explore that framework in future PRs.

**Review by BoxiLi — APPROVED** [review]

**Comment by BoxiLi** [comment]
Congrats @EnBr55 ! Feel free to open further PRs for more documentation examples as well as further development!

@nathanshammah Pinging Nathan to share the moment.

**Comment by EnBr55** [comment]
Thanks so much for all the help and advice @BoxiLi @nathanshammah !! :)

**Pull request merged**
